### PR TITLE
Base precompile code between connectors

### PR DESCRIPTION
**Pull request by mfornet**
Add precompiles, so eth-connector and erc20-connector can work from it.

### CODE REVIEW

**Review by sept-en — APPROVED** [review]
Added some comments (we already discussed some of them on the call with you), but I'm fine if we don't address them in the current PR to speed up the process. Noted here just for a future reference if this won't be fixed right now.

Looks good to me, great job!

File: src/sdk.rs [hunk 1]
```diff
@@ -278,6 +278,26 @@ pub fn predecessor_account_id() -> Vec<u8> {
...
+
+#[allow(dead_code)]
+pub fn signer_account_pk() -> Vec<u8> {
+    unsafe {
+        exports::signer_account_pk(1);
+        let bytes: Vec<u8> = vec![0u8; exports::register_len(1) as usize];
+        exports::read_register(1, bytes.as_ptr() as *const u64 as u64);
+        bytes
+    }
+}
```
[the remark this answers is not in the document]
Comment (mfornet) [reply src/sdk.rs:299]: We dont!

File: src/types.rs [hunk 2]
```diff
@@ -10,6 +11,11 @@ pub type AccountId = String;
 pub type RawAddress = [u8; 20];
 pub type RawU256 = [u8; 32]; // Little-endian large integer type.
 pub type RawH256 = [u8; 32]; // Unformatted binary data of fixed length.
+pub type Gas = u64;
+pub type Balance = u128;
+
```
Comment (mrLSD) [inline src/types.rs:16]: Just notice that we have the same type et `eth-connector`

Comment (mfornet) [reply src/types.rs:16]: Let's have it here since they are common to both

File: src/precompiles/native.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                    crate::prelude::format!(
+                        r#"{{"receiver_id": "{}", "amount": "{}", "memo": null}}"#,
+                        String::from_utf8(input.to_vec()).unwrap(),
+                        context.apparent_value.as_u128()
+                    ),
+                )
+            } else {
+                // TODO: Which is the expected result in case of failure
+                //  Stopped vs Returned vs Suicided
+                return Ok((ExitSucceed::Stopped, Vec::new(), 0));
```
Comment (mfornet) [inline src/precompiles/native.rs:108]: What is the expected PrecompileResult if it should fail?

Should we return a flag instead `1` on success, and `0` on failure?
```
return Ok((ExitSucceed::Returned, [flag].to_vec(), 0));
```

Comment (mfornet) [reply src/precompiles/native.rs:108]: From what I remember using `sha` precompile, when you called it, the result is stored in the range from [0, 32]
However, it also returns a boolean value, whether it succeeded or not. How can I return this boolean value from here?

Comment (sept-en) [reply src/precompiles/native.rs:108]: I think we should use `ExitError` for that purpose

File: src/precompiles/native.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    pub(super) const FT_TRANSFER_GAS: Gas = 100_000_000_000_000;
+
+    // TODO(#51): Determine the correct amount of gas
+    pub(super) const WITHDRAWAL_GAS: Gas = 100_000_000_000_000;
+}
+
+/// Get the current nep141 token associated with the current erc20 token.
+/// This will fail is none is associated.
+fn get_nep141_from_erc20(_erc20_token: &[u8]) -> AccountId {
+    // TODO:
```
Comment (mrLSD) [inline src/precompiles/native.rs:26]: Where it will TODO? If not planned now we should create an issue.

File: src/precompiles/native.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                (
+                    nep141_address,
+                    crate::prelude::format!(
+                        r#"{{"amount": "{}", "recipient": "{}"}}"#,
+                        amount,
+                        eth_recipient
+                    ),
+                )
+            } else {
+                // TODO: Which is the expected result in case of failure
```
Comment (mrLSD) [inline src/precompiles/native.rs:230]: Maybe for that cases, we can add log message?

File: src/precompiles/native.rs (hunk 4 above)
Comment (mfornet) [reply src/precompiles/native.rs:26]: it is done in my pr, didnt remove for it to compile

**Comment by artob** [comment]
@mfornet Is this ready to end up on chain on MainNet tomorrow? If not, please retarget it at the `develop` branch instead of `master`.

**Pull request merged**